modules/core/src/main/java/org/apache/ignite/cache/affinity/fair/FairAffinityFunction.java

```diff
@@ -865,16 +896,53 @@ else if (force) {
          * @param tier Tier.
          * @param node Node.
          * @param allowNeighbors Allow neighbors.
+         * @return Can partition assign to node.
          */
         private boolean isAssignable(int part, int tier, final ClusterNode node, boolean allowNeighbors) {
             if (containsPartition(part, node))
                 return false;
 
             if (exclNeighbors)
                 return allowNeighbors || !neighborsContainPartition(node, part);
-            else if (backupFilter == null)
-                return true;
-            else {
+            else if (affinityBackupFilter != null) {
+                List<ClusterNode> assigment = assignments.get(part);
+
+                List<ClusterNode> newAssignment;
+
+                if (tier == 0) {
+                    for (int t=1; t<assigment.size(); t++) {
+                        newAssignment = new ArrayList<>();
```

Comment:
Provide initial capacity for the new ArrayList. It can be calculated in advance.